src/Brahma.FSharp.OpenCL.Core/GPU.fs

```diff
@@ -0,0 +1,218 @@
+                                                                 System.IntPtr(src.Length * elementSize), dst, 0u, null, eventID)
+
+                                //printfn "%A" (Cl.Exception error)
+                                if error <> ErrorCode.Success
+                                then tryReplay a.ReplyChannel (Error ((Cl.Exception error):> Exception)) (Some queue)
+                                dst
+
+                            let res = read a.Source a.Destination
+                            tryReplay a.ReplyChannel (Ok res) (Some queue)
+                            0
+                }
+
+    member private this.HandleRun (queue, run:RunCrate) =
+        run.Apply
+                {
+                    new RunCrateEvaluator<int>
+                    with member this.Eval (a) =
+                            
+                            let range = a.Kernel.Range
+                            let workDim = uint32 range.Dimensions
+                            let eventID = ref Unchecked.defaultof<Event>
+                            let error =
+                                Cl.EnqueueNDRangeKernel(queue, a.Kernel.ClKernel, workDim, null,
+                                                        range.GlobalWorkSize, range.LocalWorkSize, 0u, null, eventID)
+                            
+                            if error <> ErrorCode.Success
+                            then 
+                                try 
+                                    let e = (Cl.Exception error) :> Exception
+                                    tryReplay a.ReplyChannel (Error e) (Some queue)
+                                    raise e
+                                with 
+                                | e -> printfn "%A" e
+                            else tryReplay a.ReplyChannel (Ok()) (Some queue)
+                            
+                            0
+                }
+
+    member this.GetNewProcessor () = MailboxProcessor.Start(fun inbox ->
+
+        let commandQueue = createNewCommandQueue()
+
+        let mutable itIsFirstNonqueueMsg = true
+
+        printfn "MB is started"
+        
+        let rec loop i = async {
+            let! msg = inbox.Receive()
+            match msg with
+            | MsgToHost a ->
+                //printfn "ToHost"
+                itIsFirstNonqueueMsg  <- true
+                this.HandleToHost(commandQueue, a) |> ignore
+
+            | MsgToGPU a ->
+                //printfn "ToGPU"
+                itIsFirstNonqueueMsg  <- true
+                this.HandleToGPU(commandQueue, a) |> ignore
+
+            | MsgRun a ->
+                //printfn "Run"
+                itIsFirstNonqueueMsg  <- true
+                this.HandleRun(commandQueue, a) |> ignore
+
+            | MsgFree a ->
+                //printfn "Free"
+                if itIsFirstNonqueueMsg 
+                then                    
+                    OpenCL.Net.Cl.Finish(commandQueue)
```

Comment:
Am I right that running multiple long-lasting processes in parallel (in different agents) can block almost all threads in thread pool (by calling Free at the end of each process)?

Reply:
Only if you use ```PostAndReply``` to send a ```Free``` message.  ```Finish``` blocks GPU-associated thread only. But yes, if you have a huge number of GPU-associated threads, you can block them all by sending a ```Free```messages to them simultaneously. In any case, Free is a blocking operation, because all previous commands should be finished before.